tests/butlerProxy.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python
+
+#
+# LSST Data Management System
+# Copyright 2008, 2009, 2010 LSST Corporation.
+#
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+
+from __future__ import absolute_import
+
+import os
+import shutil
+import unittest
+import lsst.utils.tests
+
+import lsst.daf.persistence as dafPersist
+import lsst.daf.base as dafBase
+
+from _testLib import isValidDateTime
+
+# Define the root of the tests relative to this file
+ROOT = os.path.abspath(os.path.dirname(__file__))
+
+
+class ButlerProxyTestCase(unittest.TestCase):
+    """A test case for the data butler finding a Mapper in a root"""
+
+    def setUp(self):
+        if os.path.exists(os.path.join(ROOT, 'root/out')):
+            shutil.rmtree(os.path.join(ROOT, 'root/out'))
+
+        self.butler = dafPersist.Butler(os.path.join(ROOT, "root"), outPath="out")
+
+    def tearDown(self):
+        del self.butler
+        if os.path.exists(os.path.join(ROOT, 'root/out')):
+            shutil.rmtree(os.path.join(ROOT, 'root/out'))
+
+    def testCheckProxy(self):
+        """Attempt to cycle a DateTime object through the butler
+        """
+        dt = dafBase.DateTime.now()
+        self.butler.put(dt, "dt", ccd=1)
+
+        # First try with immediate read, this should obviously work
+        dt = self.butler.get("dt", ccd=1, immediate=True)
+        self.assertIsInstance(dt, dafBase.DateTime)
+        self.assertTrue(isValidDateTime(dt))
+
+        # Now try again with lazy read
+        dt = self.butler.get("dt", ccd=1, immediate=False)
+        self.assertIsInstance(dt, dafPersist.readProxy.ReadProxy)
+        self.assertTrue(isValidDateTime(dt))
+
```

Comment:
It might be good to try passing something that isn't a valid `DateTime` or a proxy to `isValidDateTime` to make sure the exception type (at least) and hopefully the message are what they would be if we didn't register a proxy for it.